train.py

```diff
@@ -0,0 +1,230 @@
+import io
+import os
+
+import hydra
+import torch
+import torch.nn as nn
+from PIL import Image
+from tqdm import tqdm
+import torch.optim as optim
+import matplotlib.pyplot as plt
+from torch.utils.data import DataLoader
+from omegaconf import OmegaConf, DictConfig
+
+import wandb
+from utils.dataloaders import set_seed, create_dataloader
+from model.dcgan import Generator, Discriminator, weights_init
+
+
+def create_folders(cfg: DictConfig):
+    logger_dict = cfg.logger
+
+    for _, value in logger_dict.items():
+        if value[-1] == "/":
+            value = value[:-1]
+            os.makedirs(value, exist_ok=True)
+
+
+@torch.no_grad()
+def visualize_training(generator: Generator, fixed_noise: torch.Tensor, epoch: int, batch_idx: int, chart_path: str = "tmp/"):
+    generator.eval()
+
+    # Generate fake data using the generator and fixed noise
+    fake_data = generator(fixed_noise).detach().cpu().numpy()
+
+    # Create a figure and a grid of subplots
+    fig, axes = plt.subplots(
+        nrows=4,
+        ncols=1,
+        figsize=[10, 4],
+        gridspec_kw={"hspace": 0},
+    )
+
+    # Loop through each subplot to plot the 2-channel data
+    for i in range(len(fixed_noise)):
+        axes[i].plot(fake_data[i, 0, :])
+
+    fig.suptitle(f"Epoch {epoch}, Batch {batch_idx}")
+    fig.tight_layout()
+    # save the figure to img/ folder
+    fig.savefig(f"{chart_path}epoch_{epoch}_batch_{batch_idx}.png")
+    return fig
+
+
+def save_progress_gif(fig_list, chart_path: str = "tmp/"):
+    imgs = []
+    for fig in fig_list:
+        buf = io.BytesIO()
+        fig.savefig(buf, format="png")
+        buf.seek(0)
+        img = Image.open(buf)
+        imgs.append(img)
+
+    # Save as an animated GIF
+    imgs[0].save(f"{chart_path}progress.gif", save_all=True, append_images=imgs[1:], loop=0, duration=500)
+
+
+def train_step(
+    generator: Generator,
+    discriminator: Discriminator,
+    train_loader: DataLoader,
+    gen_optimizer: torch.optim,
+    disc_optimizer: torch.optim,
+    criterion: nn.Module,
+    cfg: DictConfig,
+    fixed_noise: torch.Tensor,
+    epoch: int,
+) -> None:
+    generator.train()
+    discriminator.train()
+
+    def random_labels(batch_size, start, end, device):
+        return (end - start) * torch.rand(batch_size, device=device) + start
+
+    fig_list = []
+
+    progress_bar = tqdm(enumerate(train_loader), total=len(train_loader))
+    for batch_idx, batch in progress_bar:
+        real_data = batch[0].to(cfg.system.device)
+        batch_size = real_data.size(0)
+
+        real_labels = random_labels(batch_size, 0.8, 1, cfg.system.device)
+        fake_labels = random_labels(batch_size, 0.0, 0.1, cfg.system.device)
```

Comment:
I'm guessing that's one of the gan training tricks? Can you add a comment explaining where this comes from? Just a link to a paper/article/repo is fine 👍

Reply:
I tried to put most of the materials I was using in README, this part is from GANhacks but I saw some papers that were talking about it in the context of images.

Reply:
You can also like a code comment to give the reader a faster path to understanding, something like this would be enough:

```python
# Trick 13 from: https://github.com/soumith/ganhacks
real_labels = random_labels(batch_size, 0.8, 1, cfg.system.device)
fake_labels = random_labels(batch_size, 0.0, 0.1, cfg.system.device)
```